crates/red_knot_python_semantic/src/types.rs

```diff
@@ -769,6 +769,14 @@ impl<'db> Type<'db> {
                 .iter()
                 .any(|&elem_ty| self.is_subtype_of(db, elem_ty)),
 
+            // `object` is the only type that can be known to be a supertype of any intersection,
+            // even an intersection with no positive elements
+            (Type::Intersection(_), Type::Instance(InstanceType { class }))
+                if class.is_known(db, KnownClass::Object) =>
+            {
+                true
+            }
+
```

Comment:
Subtype-of-object really does have to be a special case for negation types (that is, intersections with no positive elements). They are subtypes of `object` simply because `object` is the top type, so they must be, but there is no other type we can establish them to be a subtype of (other than another negation type, which is handled in the next case below.)

This case could be generalized to apply to any `self` type, not just intersections, and maybe it should be. I kind of like the fact that in most cases we don't need to special case `object` to get the right behavior (and the property test suggests we don't need to special case it for anything else). But practically speaking it might be faster to always shortcut for it.

Reply:
Conceptually it feels more elegant to me to fallback to the supertype where possible and let the handling for `object` fall out naturally for everything except this branch. But obviously if it is more performant, we should do the more performant thing.

Reply:
> This case could be generalized to apply to any `self` type, not just intersections, and maybe it should be.

We did actually do this before https://github.com/astral-sh/ruff/pull/14924, but I took it out as part of that PR because (so I thought) we didn't need it anymore

Reply:
I'm not going to generalize it in this PR; we can explore it as a possible perf optimization later.